### PR TITLE
Add GCP VM auto-discovery

### PR DESCRIPTION
**Pull request by atburke**
This PR adds node auto-discovery for GCP VMs. Part of #22815.

Depends on #28561.

### CODE REVIEW

File: lib/srv/discovery/discovery.go [hunk 1]
```diff
@@ -603,6 +631,84 @@ func (s *Server) handleAzureDiscovery() {
...
+}
+
+func (s *Server) handleGCPInstances(instances *server.GCPInstances) error {
+	client, err := s.Clients.GetGCPInstancesClient(s.ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.filterExistingGCPNodes(instances)
+	if len(instances.Instances) == 0 {
+		return trace.Wrap(errNoInstances)
```
Comment (smallinsky) [inline lib/srv/discovery/discovery.go:667]:
```suggestion
		return trace.NotFound(errNoInstances)
```

File: lib/cloud/gcp/vm.go [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+	}
+	if len(zone) == 0 {
+		return nil, trace.BadParameter("location must be set")
+	}
+
+	var instances []*Instance
+	var getInstances func() ([]*computepb.Instance, error)
+
+	if zone == types.Wildcard {
+		it := clt.InstanceClient.AggregatedList(ctx, &computepb.AggregatedListInstancesRequest{
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/cloud/gcp/vm.go:177]: https://github.com/gravitational/teleport/pull/28561#discussion_r1263156493

File: lib/cloud/gcp/vm.go [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+func (clt *instancesClient) ListInstances(ctx context.Context, projectID, zone string) ([]*Instance, error) {
+	if len(projectID) == 0 {
+		return nil, trace.BadParameter("projectID must be set")
+	}
+	if len(zone) == 0 {
+		return nil, trace.BadParameter("location must be set")
+	}
+
+	var instances []*Instance
+	var getInstances func() ([]*computepb.Instance, error)
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/cloud/gcp/vm.go:174]: Having the closure here lets us unify the return types of `AggregatedList` and `List`. I think if we moved it outside, we'd have to have two functions and either extra `if`s or a closure to unify them, which would defeat the point of moving them out.

File: lib/srv/discovery/discovery.go (hunk 1 above)
Comment (atburke) [reply lib/srv/discovery/discovery.go:667]: `errNoInstances` is already an error type, so we can't wrap it with `trace.NotFound`.

**Comment by atburke** [comment]
FYI @smallinsky this PR contains changes that are also in #28561 (mostly lib/cloud/gcp/*), so I've addressed your comments on those files over there.

**Comment by atburke** [comment]
Friendly ping @smallinsky

File: lib/services/matchers.go [hunk 4]
```diff
@@ -353,10 +353,13 @@ var SupportedAzureMatchers = []string{
 const (
 	// GCPMatcherKubernetes is the GCP matcher type for GCP kubernetes.
 	GCPMatcherKubernetes = "gke"
+	// GCPMatcherVM is the GCP matcher for GCP VMs.
+	GCPMatcherVM = "vm"
```
Comment (smallinsky) [inline lib/services/matchers.go:357]: vm is quite broad term. What would you say about replacing  this matcher name with `gce` - Google Compute Engine ?

**Review by smallinsky — APPROVED** [review]

**Review by xacrimon — APPROVED** [review]

**Pull request merged**